Preserve entitlements in MacOS signer

jtschuster opened the pull request:
Preserve entitlements in the manages Mach-O signer for both regular apphost and singlefilehost.

Adds entitlements to singlefilehost during the build.

Refactors the signature blob types to be classes that implement IBlob. The various named blobs derive from SimpleBlob or SuperBlob (except for CodeDirectoryBlob which is more complex). SimpleBlob and SuperBlob provide methods for accessing info about the blobs and writing to a file. IMachFile is used for writing to allow reading and writing from/to a Stream or MemoryMappedViewAccessor.

The CodeSignature class was replaced with just the EmbeddedSignatureBlob, a SuperBlob containing other relevant signature blobs. Methods from the CodeSignature class were move to the relevant blob class.

HostWriter.SetAsBundle was moved to the Bundle class to avoid having to remap the memory mapped file multiple times. Since the host destination stream is never closed between creating the bundle and SetAsBundle / Signing the bundle, we don't need to create a new copy with a new inode to sign and replace the existing bundle.

cc @dotnet/appmodel

jtschuster committed:
Use managed signer for bundles

jtschuster committed:
Fix tests and write out symtab

- Write out the updated symtab command when writing Mach-O file
- Pass macosCodesign through to test CreateAppHost methods
- Remove redundant `codesign` checks
- Warn when bundler is told to sign the bundle for a non-macos target

jtschuster committed:
Fix test failures

- Allow macosCodesign to be true for non-mac bundles
- Allow some padding between string table and signature

jtschuster committed:
Flip comparison

jtschuster committed:
Add comment to TryAdjustHeadersForBundle, and throw if it returns false

jtschuster committed:
Don't memory-map file if it's not a Mach-O, remove unused file.

jtschuster committed:
Fix typo, restructure IsSigned check.

jtschuster committed:
PR Feedback: Fomatting and comments

jtschuster committed:
PR feedback:

- Use the same memory-mapped file instance for placeholder replacement
  and signing
- formatting changes

jtschuster committed:
Create new file for macos signing, update LINKEDIT.VMSize to fit signature

jtschuster committed:
Merge branch 'main' into ManagedSignBundles

jtschuster committed:
Merge branch 'main' into ManagedSignBundles

jtschuster committed:
PR feedback

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into ManagedSignBundles

jtschuster committed:
Remove old file that's not needed

jtschuster committed:
Merge branch 'ManagedSignBundles' of https://github.com/jtschuster/runtime into ManagedSignBundles

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into ManagedSignBundles

jtschuster committed:
Use constant for alignments

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into ManagedSignBundles

jtschuster committed:
Wip

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into ManagedSignBundles

jtschuster committed:
revert dotnet.sh

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into ManagedSignBundles

jtschuster committed:
Wip

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into ManagedSignBundles

jtschuster committed:
Ensure entitlements are preserved

jtschuster committed:
Update doc comments

jtschuster committed:
Add entitlements to singlefilehost

jtschuster committed:
Check for identical code signatures in tests

jtschuster committed:
undo test throw

jtschuster committed:
Remove CodeSignature class use EmbeddedSignatureBlob instead

- Delete CodeSignature class and move methods on relevant blobs instead.
- Use EmbeddedSignatureBlob as the "signature" for the MachO file.
- Save old EmbeddedSignatureBlob in Bundling to preserve the entitlements.
- Add test to verify that the entitlements are preserved in the bundle.

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into PreserveEntitlements

jtschuster committed:
Remove CodeSignature class, unsign before enabling testonly behaviors

jtschuster committed:
Rename file to match class, don't set length of file while mapped

agocke commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:
```diff
@@ -273,80 +274,149 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
             long headerOffset = 0;
-            using (FileStream bundle = File.Open(bundlePath, FileMode.Open, FileAccess.ReadWrite))
-            using (BinaryWriter writer = new BinaryWriter(bundle, Encoding.Default, leaveOpen: true))
+            string destinationDirectory = new FileInfo(bundlePath).Directory?.FullName ?? "";
```
The `?? ""` here looks suspicious. I don't see how that could be correct.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:
```diff
@@ -273,80 +274,149 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
+                using (FileStream hostSourceStream = File.OpenRead(hostSource))
                 {
-                    MachObjectFile.RemoveCodeSignatureIfPresent(bundle);
+                    hostSourceStream.CopyTo(bundleStream);
```
Is there any benefit to this being async? My guess is that MSBuild doesn't have an async entry point, but I could be wrong about that.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:
```diff
@@ -273,80 +274,149 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
-
-                    if (IsHost(relativePath))
+                    long? newLength = null;
+                    using (MemoryMappedFile bundleMap = MemoryMappedFile.CreateFromFile(bundleStream, null, bundleStream.Length, MemoryMappedFileAccess.ReadWrite, HandleInheritability.None, leaveOpen: true))
```
It feels weird to copy the file, then memory map the copied file.

If I remember correctly, the memory map API supports reading the input file, modifying the input, then writing the modifications to an output file.

jtschuster replied:
Ah, yeah we should be able to do that.

jtschuster replied:
Updated to create an in-memory memory-mapped file, write all the bundle data, then write out to the output path.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:
```diff
@@ -273,80 +274,149 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
+                        bundleStream.SetLength(newLength.Value);
                     }
+                }
+                using (BinaryWriter writer = new BinaryWriter(bundleStream, Encoding.Default, leaveOpen: true))
```
Could we also do this with memory-mapping, rather than re-opening the written file?

jtschuster replied:
Yeah, we probably could calculate the size of the files ahead of time and extend the file to that length before memory mapping.

jtschuster replied:
Updated to use a single memory-mapped file rather than memory mapping multiple times.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:
```diff
@@ -273,80 +274,149 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
+                                                BitConverter.GetBytes(headerOffset),
+                                                pad0s: false);
+
+                    if (_target.IsOSX)
```
Seems strange to me that this bit and the codesign header removal are in separate places? Shouldn't they be in the same place?

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/BinaryFormat/Blobs/Blob.cs:
```diff
@@ -0,0 +1,88 @@
+/// This class implements the basic functionality of reading and writing the magic number and size of the blob.
+/// All blobs are expected to be in big-endian format.
+/// </summary>
+internal abstract class Blob
```
I'm suspicious of the use of inheritance, rather than interface implementation, here. It doesn't seem like it saves much code, and it creates complex inheritance relationships where some behavior is provided in the base class and some behavior is provided in the derived class. Moreover, there are interdependencies that can go out of sync -- for instance, `Size` is used in `Write` and `GetBytes`, so if you override `Write` but don't override `Size`, you get the wrong results in `GetBytes`.

It seems simpler and less error-prone to define an interface and force every implementor to write out the implementations that are correct for their blob.

elinor-fung replied:
This is the part that raises a flag for me (from the PR description):

> Each class in the hierarchy is responsible for their fields only, and derived classes use the base.Size property to determine the offset of new fields in the blob.

Aside from the override of the property itself, once the derived needs to to reason about whether or not it needs to use its implementation or the base one, I think either the relationship between base and derived should be redefined or the base functionality should just helpers instead of a base class.

jtschuster replied:
Restructured to avoid reaching into the base class for implementations.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/BinaryFormat/Blobs/SimpleBlob.cs:
```diff
@@ -0,0 +1,49 @@
+        base.Write(accessor, offset);
+        if (_data.Length != 0)
+        {
+            accessor.WriteArray(offset + base.Size, _data, 0, _data.Length);
```
It feels very suspicious that you're reaching into the base type implementation here. I think use of `base` outside of an override of the same method is almost always a layering violation.

jtschuster replied:
Removed this

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/BinaryFormat/Blobs/SimpleBlob.cs:
```diff
@@ -0,0 +1,49 @@
+
+    public SimpleBlob(MemoryMappedViewAccessor accessor, long offset) : base(accessor, offset)
+    {
+        accessor.Read(offset + sizeof(uint), out uint size);
```
Doing I/O in a constructor is generally bad practice. Constructors can't return errors or be async, so you heavily narrow your options. Most users will also not expect constructors to do significant work.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/BinaryFormat/Blobs/EmbeddedSignatureBlob.cs:
```diff
@@ -0,0 +1,248 @@
+/// but will be present in newly created signatures.
+/// Optionally, it may also contain the EntitlementsBlob and DerEntitlementsBlob.
+/// </summary>
+internal sealed class EmbeddedSignatureBlob : SuperBlob
```
This doesn't look like an inheritance relationship to me, it looks like composition. Based on the linked code above, an `EmbeddedSignatureBlob` _contains_ a superblob, but is not a "kind" of superblob. I think this is true of all superblobs.

agocke replied:
So is the only purpose setting the MagicField number? That still seems like overkill for inheritance. Couldn't this be an enum value that's passed in?

jtschuster replied:
I just found it helpful to have different type names for each blob. Using inheritance for implementing the Magic enum just made sense after that decision. Is there that much overhead to having the different classes?

jtschuster committed:
Update bundler to memory map a file once, add test for inode change when signing

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/MachO/BinaryFormat/Blobs/SuperBlob.cs:
```diff
@@ -0,0 +1,140 @@
+/// This class handles reading and writing of all the sub-blobs.
+/// The blob contains the following structure:
+/// </summary>
+internal abstract unsafe class SuperBlob : Blob
```
Yes, the EmbeddedSignatureBlob includes the RequirementsBlob which are both SuperBlobs.

jtschuster committed:
Don't abuse inheritance, add Stream file abstraction

- Don't use multiple levels of inheritance for blobs
- Add IMachOFileReader/Writer to abstract Stream and MemoryMappedViewAccessor based files
- Use IMachOFileReader/Writer as the interface for reading/writing to the disk for MachO files and blobs

jtschuster committed:
Update reading BigEndian values

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into PreserveEntitlements

jtschuster committed:
Use composition + interfaces instead of inheritance for blobs

jtschuster committed:
Merge branch 'main' into PreserveEntitlements

jtschuster commented:
Superceded by https://github.com/dotnet/runtime/pull/116566
